apps/server/lib/lexical/server/code_intelligence/completion/translations/macro.ex

```diff
@@ -596,6 +568,57 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.Macro do
     :skip
   end
 
+  defp function_snippet(kind, label, %Candidate.Macro{} = macro, builder, env) do
+    label = "#{macro.name} (#{label})"
+
+    snippet =
+      with %Position{} = position <- Env.prev_significant_position(env),
+           {:ok, name, arity} <- extract_spec_name_arity(env, position) do
+        args_snippet =
+          if arity == 0 do
+            ""
+          else
+            "(" <> Enum.map_join(1..arity, ", ", &"${#{&1}:arg_#{&1}}") <> ")"
+          end
+
+        """
+        #{kind} #{name}#{args_snippet} do
+          $0
+        end
+        """
+      else
+        _ ->
+          """
+          #{kind} ${1:name}($2) do
+            $0
+          end
+          """
+      end
+
+    env
+    |> builder.snippet(snippet,
+      detail: macro.spec,
+      kind: :class,
+      label: label,
+      filter_text: macro.name
+    )
+    |> builder.set_sort_scope(SortScope.global())
+  end
+
+  defp extract_spec_name_arity(%Env{} = env, %Position{} = position) do
+    with {:ok, [maybe_spec | _]} <- Ast.path_at(env.analysis, position),
+         {:@, _, [{:spec, _, [typespec]}]} <- maybe_spec,
+         {:"::", _, [{name, _, args}, _return]} <- typespec do
+      if is_list(args) do
+        {:ok, name, length(args)}
+      else
+        {:ok, name, 0}
+      end
+    else
+      _ -> :error
+    end
+  end
```

Comment:
should this go into an Ast.Typespec module?

Reply:
Maybe? I think Ast.Typespec might be too specific, since some of the logic here could/would be shared for similar, non-spec AST. e.g. getting the name and arity for any call (we do this for entity resolution). I can't think of a name just yet, but I'm thinking a more general Ast.Something with functions that can be easily composed using `with` (fetch semantics). So this particular function would end up something like

```elixir
with {:ok, spec_ast} <- Something.typespec_at(analysis, position),
     {:ok, call} <- Something.typespec_call(spec_ast) do
  Something.call_name_and_arity(call)
end
```

I think that having these in one module will prevent having to have *another* module for them to share helper functions and will make them easier to find.

Reply:
While I was trying to come up with a name for this module, it dawned on me that the correct name is already there: `Ast.Detection`.

What if we expanded the detection behaviour to include `detect` as well as `detected?`. It would return `{:ok, node}` or `:error`. It might make the most sense as an optional callback, since there are some detectable things that don't necessarily have a node that encapsulates them, e.g. comments.

Reply:
Love the name.

Reply:
Discussed on Discord; going to save this for another PR.